### ISSUE TITLE
Using add_blank_wires does not preserve the Connectivity

### ISSUE THREAD

**Issue by nquetschlich** [CQCL/tket#777]
Hi,

this issue is closely related to #770 (and #774). As mentioned in #770, I tried to compile quantum circuit to certain target devices with the goal of having a resulting quantum circuit with as many qubits as the targeted device provides. @cqc-alec suggested to use `add_blank_wires` which I have used after the actual compilation has finished (in this case, for a coupling map which is directed): 

```python
from pytket import architecture, OpType
from pytket.qasm import circuit_from_qasm_str

from pytket.passes import (
    auto_rebase_pass,
    CXMappingPass
)
from pytket.placement import GraphPlacement
from pytket.qasm import circuit_from_qasm_str

qc_tket = circuit_from_qasm_str('OPENQASM 2.0;\ninclude "qelib1.inc";\nqreg q[3];\ncreg c[2];\nh q[0];\nh q[1];\nx q[1];\nx q[2];\nh q[2];\ncx q[0],q[2];\nh q[0];\ncx q[1],q[2];\nx q[1];\nh q[1];\nbarrier q[0],q[1],q[2];\nmeasure q[0] -> c[0];\nmeasure q[1] -> c[1];\n')

cmap = [[0, 1], [0, 7], [1, 2], [2, 3], [7, 6], [6, 5], [4, 3], [4, 5]]
arch = architecture.Architecture(cmap)
native_gate_set_rebase = auto_rebase_pass({OpType.Rz, OpType.SX, OpType.X, OpType.ECR, OpType.Measure})

native_gate_set_rebase.apply(qc_tket)
placer = GraphPlacement(arch) 
CXMappingPass(arc=arch, placer=placer, directed_cx=True, delay_measures=False).apply(qc_tket)
print(qc_tket.valid_connectivity(arch, directed=True))

diff = 8 - qc_tket.n_qubits 
qc_tket.add_blank_wires(diff)

print(qc_tket.valid_connectivity(arch, directed=True))
```

While the first `print` statement returns `True` and, by that, confirm the valid connectivity of the compiled circuit, the second `print` statement returns `False` after adding the blank wires. To me, this is was unexpected and I wanted to confirm whether this behavior is intended.

**Comment by nquetschlich** [comment]
> Hi,
> 
> The `valid_connectivity` method also checks whether the names of the qubits in your circuit match the nodes in the architecture.
> 
> In this case, `arch = architecture.Architecture(cmap)` constructs an `Architecture` with nodes: "node[0]", "node[1]", ..., "node[7]" ("node" is the default name for [Nodes](https://cqcl.github.io/tket/pytket/api/circuit.html#pytket.circuit.Node)).
> 
> The default qubits in your circuit are "q[0]", "q[1]", "q[2]". The `CXMappingPass` pass then routes the circuit and renames the qubits to "node[0]" , "node[1]", and "node[2]". You can verify this with `qc_tket.qubits`.
> 
> Then when you call `add_blank_wires()` it actually adds "q[0]", "q[1]", "q[2]", "q[3]", "q[4]" to your circuit. Since these new qubits are not in `arch ` the `valid_connectivity` returns `False`.
> 
> The easiest solution is to add the qubits before routing so `CXMappingPass` renames all qubits.
> 
> ```python
> qc_tket = circuit_from_qasm_str('OPENQASM 2.0;\ninclude "qelib1.inc";\nqreg q[3];\ncreg c[2];\nh q[0];\nh q[1];\nx q[1];\nx q[2];\nh q[2];\ncx q[0],q[2];\nh q[0];\ncx q[1],q[2];\nx q[1];\nh q[1];\nbarrier q[0],q[1],q[2];\nmeasure q[0] -> c[0];\nmeasure q[1] -> c[1];\n')
> diff = 8 - qc_tket.n_qubits 
> qc_tket.add_blank_wires(diff)
> 
> cmap = [[0, 1], [0, 7], [1, 2], [2, 3], [7, 6], [6, 5], [4, 3], [4, 5]]
> arch = architecture.Architecture(cmap)
> native_gate_set_rebase = auto_rebase_pass({OpType.Rz, OpType.SX, OpType.X, OpType.ECR, OpType.Measure})
> 
> native_gate_set_rebase.apply(qc_tket)
> placer = GraphPlacement(arch) 
> CXMappingPass(arc=arch, placer=placer, directed_cx=True, delay_measures=False).apply(qc_tket)
> print(qc_tket.valid_connectivity(arch, directed=True))
> ```

I see, thank you very much for the clarification.

**Comment by yao-cqc** [comment]
Thanks for reporting this!  As commented on #776, ``FullPeepholeOptimise`` doesn't preserve connectivity so that's why the `if` condition evaluates `True`. The segfault is caused by a bug in the routing code that has just been fixed by #840.

**Comment by nquetschlich** [comment]
Thanks a lot for clarification. Since I posted my ongoing problems in #776, I will close this issue.
